### PR TITLE
esn-frontend-calendar#474: Added API to get the ics of a calendar

### PR DESCRIPTION
**Pull request by huy-ta**
Resolves https://github.com/OpenPaaS-Suite/esn-frontend-calendar/issues/474.

The tests for the `getCalendarIcs` function should be written in https://github.com/OpenPaaS-Suite/esn-frontend-calendar/issues/475.

Example response:

```js
[
  {
    href: '/calendars/5f60334e78d1b021351f9f6e/inbox/sabredav-c3fa55f7-1c62-413c-9be4-438a1f4d48a4.ics',
    etag: '"a649ba6a6a35aeceb7f23b066ecb43e8"',
    ics: 'BEGIN:VCALENDAR\n' +
      'VERSION:2.0\n' +
      'PRODID:-//Sabre//Sabre VObject 4.1.3//EN\n' +
      'CALSCALE:GREGORIAN\n' +
      'METHOD:REQUEST\n' +
      'BEGIN:VTIMEZONE\n' +
      'TZID:Asia/Jakarta\n' +
      'BEGIN:STANDARD\n' +
      'TZOFFSETFROM:+0700\n' +
      'TZOFFSETTO:+0700\n' +
      'TZNAME:WIB\n' +
      'DTSTART:19700101T000000\n' +
      'END:STANDARD\n' +
      'END:VTIMEZONE\n' +
      'BEGIN:VEVENT\n' +
      'UID:a101a7ee-b4e8-4c79-aa2c-1e1516ceb3ae\n' +
      'TRANSP:OPAQUE\n' +
      'DTSTART;TZID=Asia/Saigon:20210318T120000\n' +
      'DTEND;TZID=Asia/Saigon:20210318T123000\n' +
      'CLASS:PUBLIC\n' +
      'X-OPENPAAS-VIDEOCONFERENCE:\n' +
      'SUMMARY:Invite admin\n' +
      'DESCRIPTION:12312313123\n' +
      'ORGANIZER;CN=John10 Doe10:mailto:user10@open-paas.org\n' +
      'ATTENDEE;PARTSTAT=NEEDS-ACTION;RSVP=TRUE;ROLE=REQ-PARTICIPANT;CUTYPE=INDIVI\n' +
      ' DUAL;CN=admin admin:mailto:admin@open-paas.org\n' +
      'ATTENDEE;PARTSTAT=ACCEPTED;RSVP=FALSE;ROLE=CHAIR;CUTYPE=INDIVIDUAL:mailto:u\n' +
      ' ser10@open-paas.org\n' +
      'DTSTAMP:20210319T044228Z\n' +
      'SEQUENCE:0\n' +
      'END:VEVENT\n' +
      'END:VCALENDAR\n'
  }
]
```

### CODE REVIEW

**Review by rezk2ll — APPROVED** [review]

File: lib/api/calendars.ts [hunk 1]
```diff
@@ -1,10 +1,32 @@
+import convert from 'xml-js';
 import { DAVClient } from '../DAVClient';
 
 const BASE_PATH = '/calendars';
 
-export const getInbox = (client: DAVClient) => async (userId: string): Promise<string> => {
-  return client.requestText({
-    url: `${BASE_PATH}/${userId}/inbox`,
+export interface MultiStatusResponse {
```
Comment (renaudboyer) [inline lib/api/calendars.ts:6]: This doesn't need to be exported.

We may put XML interfaces in a separate file

File: test/HTTPFetchClient.test.ts [hunk 2]
```diff
@@ -48,6 +48,24 @@ describe('The HTTPFetchClient class', () => {
       expect(fetch).toHaveBeenCalledTimes(1);
       expect(fetch).toHaveBeenCalledWith(requestOptions.url, { method: requestOptions.method, headers: requestOptions.headers, body: requestOptions.body });
     });
+
+    it('should send a fetch request with the corrent params and throw an error when the response status is not ok', async (done) => {
```
Comment (renaudboyer) [inline test/HTTPFetchClient.test.ts:52]:
```suggestion
    it('should send a fetch request with the correct params and throw an error when the response status is not ok', async (done) => {
```

File: test/HTTPFetchClient.test.ts [hunk 3]
```diff
@@ -91,6 +109,24 @@ describe('The HTTPFetchClient class', () => {
       expect(fetch).toHaveBeenCalledWith(requestOptions.url, { method: requestOptions.method, headers: requestOptions.headers, body: requestOptions.body });
       expect(result).toBe(jsonBodyContent);
     });
+
+    it('should send a fetch request with the corrent params and throw an error when the response status is not ok', async (done) => {
```
Comment (renaudboyer) [inline test/HTTPFetchClient.test.ts:113]: Copy/Paste spotted :p

File: test/HTTPFetchClient.test.ts [hunk 4]
```diff
@@ -134,5 +170,23 @@ describe('The HTTPFetchClient class', () => {
       expect(fetch).toHaveBeenCalledWith(requestOptions.url, { method: requestOptions.method, headers: requestOptions.headers, body: requestOptions.body });
       expect(result).toBe(textBodyContent);
     });
+
+    it('should send a fetch request with the corrent params and throw an error when the response status is not ok', async (done) => {
```
Comment (renaudboyer) [inline test/HTTPFetchClient.test.ts:174]: Same here

File: README.md [hunk 5]
```diff
@@ -7,15 +7,19 @@ A DAV client to send requests to DAV servers.
...
 ```javascript
-import { DAVClient, getInbox, HTTPFetchClient } from 'dav-client';
+import { DAVClient, getCalendarIcs, HTTPFetchClient } from 'dav-client';
 
 const [username, password] = ['admin@open-paas.org', 'secret'];
 const httpClient = new HTTPFetchClient();
 const davClient = new DAVClient({ baseURL: 'http://0.0.0.0:8001', httpClient, headers: { Authorization: `Basic ${Buffer.from([username, password].join(':')).toString('base64')}`} });
 
-getInbox(davClient)('5f60334e78d1b021351f9f6e')
+getCalendarIcs(davClient)('5f60334e78d1b021351f9f6e', 'inbox')
```
Comment (renaudboyer) [inline README.md:16]: I think we can let getInbox for now.

getCalendar is more generic and need more parameters in the XML REPORT to work.

Eventually getInbox may leverage a getCalendar.

Comment (huy-ta) [reply README.md:16]: OK I'll roll back to `getInbox`.

**Pull request merged**